algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+S ∈ ideals_not_univ α → ∃ (m : set α) (hm : is_ideal α m), @is_maximal_ideal _ _ m hm ∧ S ⊆ m :=
+begin
+  intro h,
+  cases h with h1 h2,
+  have z := zero_ne_one.to_maximal_ideal (α/S),
+  specialize z (λ h, h2 $ is_ideal.mem_of_coset_zero S 1 h.symm),
+  cases z with m hm,
+  cases hm with h1 h2,
+  existsi quotient_to_ideal S m,
+  existsi quotient_to_ideal.is_ideal S m,
+  have hsm : S ⊆ quotient_to_ideal S m,
+    intro x,
+    simp [is_ideal.mem_iff_coset_zero S,is_ideal.to_coset],
+    intro hx,
+    rw hx,
+    apply is_ideal.zero_mem,
+  split,
+  cases h2,
+  constructor,
+  intro h,
+  apply h2_ne_univ_ideal,
+  rw [set.set_eq_def] at h,
+  apply set.ext,
+  intro x,
+  cases is_ideal.coset_rep x with y hy,
+  rw ← hy,
+  specialize h y,
+  simp at *,
+  exact h,
+  intros T _ ht,
+  have hst : S ⊆ T := set.subset.trans hsm ht,
+  specialize h2_no_between (ideal_to_quotient S T),
+  have ht2 := ideal_to_quotient.subset S _ _ ht,
+  rw quotient_to_contains_to_quotient at ht2,
+  specialize h2_no_between ht2,
+  cases h2_no_between with h h,
+  left,
+  apply set.eq_of_subset_of_subset,
+  rw set.subset.antisymm_iff at h,
+  have h3 := quotient_to_ideal.subset S _ _ h.1,
+  rw contains_to_quotient_to_contains at h3,
+  exact h3,
+  exact hst,
+  exact ht,
+  right,
+  apply set.ext,
+  rw set.set_eq_def at h,
+  intro x,
+  specialize h (is_ideal.to_coset S x),
+  simp at *,
+  cases h with y hy,
+  cases hy with hy1 hy2,
+  calc x = y - (y - x) : by norm_num
+     ... ∈ T : is_ideal.sub_mem hy1 (hst hy2),
+  exact hsm
+end
+
+-- Corollary 1.4 end
+
+
+-- Corollary 1.5 start
+
+theorem not_unit.to_maximal_ideal {α : Type u} [comm_ring α] (x : α) (h : x ∈ nonunits α) :
+∃ (m : set α) (hm : is_ideal α m), @is_maximal_ideal _ _ m hm ∧ x ∈ m :=
+begin
+  have z := ideals_not_univ.to_maximal_ideal (principal_ideal x),
+  specialize z
+    (begin
+      split,
+      exact principal_ideal.is_ideal x,
+      intro hx,
+      have := is_ideal.univ_of_one_mem _ hx,
+      apply h,
+      apply (unit_iff_principal_ideal_eq_one x).2,
+      rw this,
+      rw principal_ideal_one_eq_univ
+    end),
+  cases z with m z,
+  cases z with hm z,
+  existsi m,
+  existsi hm,
+  split, exact z.1,
+  apply set.mem_of_mem_of_subset _ z.2,
+  exact mem_principal_ideal x
+end
+
+-- Corollary 1.5 end
+
+
+def is_local (α : Type u) [comm_ring α] : Prop :=
+∀ S T [is_ideal α S] [is_ideal α T] [@@is_maximal_ideal _ S _inst_2] [@@is_maximal_ideal _ T _inst_3], S = T
+
+theorem list.sum_singleton {α : Type u} [semiring α] (x : α) : list.sum [x] = x :=
+by simp
+
+theorem list.sum_mul {α : Type u} [semiring α] :
+∀ (L : list α) (x : α), (L.map (λ b, b * x)).sum = L.sum * x
+| []     x := by simp; rw [list.sum_nil]
+| (h::t) x := by simp [list.sum_mul t x, add_mul]
+
+section generate
+
+variables {α : Type u} [comm_ring α] (S S₁ S₂ T : set α) (x : α) [is_ideal α T]
+
+def span : set α := { x | ∃ L : list α, (∀ x ∈ L, ∃ y z, y ∈ S ∧ y * z = x) ∧ L.sum = x }
```

Comment:
Please reuse `span` in `linear_algebra`. The list construction makes it quite complicated to work with. The inductive version in `linear_algebra` is much easier to work with.